Article first draft (#135)
hassanobeid1994/tr_b_causal_2020: article/sections/_5_graph_testing.tex

\section{Testing of Causal Graphs}
\label{sec:graph-testing}

\subsection{Description}
\label{sec:testing-description}

\subsubsection{Testing observable assumptions}
\label{sec:observable-tests}
In the last section, we reviewed a process for creating an initial causal graph using expert opinion.
Critically, after drafting a causal graph, we should immediately test it against empirical data.
This is important because, if our graph captures inaccurate assumptions about the data generating process, then we have no reason to think that our conclusions from using the graph will be accurate.

> 

Reviewer: I totally agree with this, but I wonder if we would benefit from making the language less strict. 
I worry that framing the empirical testing as being so critical may intimidate people, whereas I think that going through the process of creating a graph with expert opinion is still a good step in the right direction, even without testing. 

My suggestions are for very minor edits:
- Dropping words like "immediately"
- Use a language that talks about testing as being a way to make inferences more robust and defensible. 

Below is a suggested re-write: 

In the last section, we reviewed a process for creating an initial causal graph using expert opinion.
The natural next step after creating an initial graph is to test its implications against empirical data.
This is important because, if our graph captures inaccurate assumptions about the data generating process, then we have no reason to think that our conclusions from using the graph will be accurate. 
The process of testing the causal graph therefore adds robustness and credibility to one's final conclusions.

Author: Both suggestions taken!

To test our causal graphs against data, we will first test the implications of our graph that involve observable variables only.
We will defer the task of testing implications that involve unobserved / latent variables to later in this subsection.

> For now, recall our discussion in Section \ref{sec:graph-overview} about the two basic implications of causal graphs: marginal independence and conditional independence.

Reviewer: Note to myself @hassan-obeid : 
- Make sure I emphasize marginal independence in the graph-overview section.

In both cases, direct testing of marginal or conditional independence amongst nodes in the causal graph may be difficult.
Indeed, there are no direct tests of conditional independence that can detect all types of dependence, especially for continuous variables \citep{bergsma_2004_testing, shah_2020_hardness}.

As a result of this hardness, there are a myriad of research efforts aimed at testing conditional independence of two variables $X$ and $Y$, given a third variable $Z$ and any number of assumptions about the variables or the test statistic itself.
Some researchers proceed under the assumption that one has access to an approximation of the conditional distribution of $X \mid Z$ \citep{candes_2018_panning, berrett_2019_conditional}.
Other researchers designed conditional independence tests for general cases, assuming smoothness of the underlying data distributions and assuming accurate estimation of the distribution of the test statistic under the null hypothesis of conditional independence (e.g. \citet{zhang_2012_kernel, strobl_2019_approximate}).

> Different from (but not excluding) these approaches, we will take an easier and less decisive route.

Reviewer: Can't directly suggest edits for some reason. Suggestion:
"Different from (but not excluding) these approaches, we will take an easier and less decisive route." becomes 
"In this chapter, we will take an easier and less decisive route to testing conditional independence that is influenced by those approaches."

Author: Yep, that's way clearer!

If a pair of variables have conditionally or marginally independent distributions, then their statistical moments will also be conditionally or marginally independent.
Instead of testing for marginal or conditional independence in distribution, we will perform a more tractable test for marginal or conditional independence in means.
If the variables in question are not conditionally or marginally independent in their means, then we know they are not independent in their distributions.

> Conversely, even if a set of variables are marginally or conditionally independent in their means, this \textbf{does not} imply that the variables are independent in distribution.

Reviewer: I suggest adding one more sentence saying something like "in other words, failing to reject that the variables are conditionally or marginally independent in their means adds a layer of defensibility to one's analysis, although it does not prove conditional independence".

Author: I added: "Mean independence simply provides justification for placing greater belief in the variables being distributionally independent."

Of particular importance were our simulations under the null hypothesis where $X$ was conditionally independent of $Y$.
Our initial simulations used random forests as our conditional expectation models and permutations as resampling methods.
Random forests are a non-parametric method that would allow us to have less fear of model misspecification, and permutations are easy to implement.
However, under the null hypothesis, we discovered that the tests based on the random forest models did not result in p-values that were uniformly distributed.
Given that we planned to use these procedures in a manner akin to hypothesis-testing, we hoped that our test statistics would be U-statistics.
When we switched from the combination of random forests and permutations to linear regressions and permutations, our p-values were indeed empirically, uniformly distributed.
Moreover, we still retained high power.

We do not claim that these choices for assessing mean independence will always be appropriate.
Indeed, one should assess one's tests on simulated data that resembles one's real data.

> For our illustrative purposes though, the combination of linear regressions, permutations, and $R^2$ resulted in adequate tests of marginal and conditional mean-independence.

Reviewer: This is so important and I'm glad you highlighted this process!

\subsubsection{Testing assumptions involving latent variables}
\label{sec:latent-tests}
% Describe latent variable testing
Now that we have described testing with observable variables, we can more easily describe conditional independence tests that involve unobserved (i.e., latent) variables.
Indeed, when dealing with observational data, we will frequently find ourselves not having observed all variables that are of interest.
Nevertheless, we still wish to test whether our data contradicts our graph.
One way to directly extend our conditional independence testing to account for latent variables is to adopt a missing data perspective and impute the latent variables from a prior distribution.
In particular, we can generalize our previous tests as follows.

First, we can consider that instead of performing one test with a set of observed $X$, observed $Y$ and observed $Z$, we instead perform tests of observed $X$, observed $Y$, and imputed $Z$.
This recasts the randomness underlying the null-distribution in our original test statistic of $R^2$ as a function of our permutation of $Y$ and our imputation of $Z$.

> Here, we impute $Z$ by sampling from the prior or posterior distribution of $Z$, depending on whether we're testing independencies before or after performing inference on our model's parameters.

Reviewer: I feel this subsection needs more information on the process of imputing Z. Even to me, it's still unclear how one would have a prior or posterior on Z if it's unobserved. I can understand having a prior distribution (if one theorizes on what Z is), but even that would be a marginal distribution of Z as opposed to a conditional one wrt X or Y (or a joint distribution of Z, X, Y). I can't really understand what a posterior distribution of Z means -- posterior based on what model of Z?

Instead of plotting a single line versus a distribution, we now plot two distributions against one another.
We first plot the distribution of ``observed'' test statistics that we computed using the observed $X$, observed $Y$ and imputed $Z$ values.
Then, we plot the distribution of ``sampled'' test statistics using prior samples of $\left( X, Y, Z \right)$ as a reference.
Here, we have one value per imputed vector $Z$ in both the observed and sampled distributions.
However, for the distribution of ``sampled'' test statistics, we marginalize over permutations of $Y$ since this distribution represents the null hypothesis of conditional or marginal independence.

That last paragraph may have been confusing, so we'll walk through an example.
Figure \ref{fig:deconfounder-graph} shows the causal graph implied by applying the deconfounder algorithm of \citet{wang_2019_blessings} to our dataset.
We describe this application more thoroughly in Section \ref{sec:latent-confounding}.
For now, we present the graph to highlight the assumptions that we will test.

> Specifically, we'll examine the assumption that the observed number of licensed drivers in a household is independent of the observed number of automobiles in that household, conditional on the latent variable $X^{*}$.

Reviewer: Ah, I see now. So these tests only apply when you have a model to compute Z, e.g. the deconfounder algorithm.

So in other words, assuming that you *know* the latent variable Z, can we test whether X and Y are conditionally independent. 

This strikes me as a little bit unintuitive. Particularly, here are my thoughts/concerns:
- To re-frame the problem: we want to test conditional independence, conditional on a latent variable. To do that, we need some model of the latent confounder that gives us its posterior distribution, or, if we're testing conditional independence conditional on a prior distribution, we need to assume a prior. The test is then testing for many things at once, which to me makes it hard to interpret and I'm struggling to understand its usefulness. In particular, failing to reject the null can either mean i) our model of the confounder is wrong, or ii) there are additional confounders that are not represented. I noticed you touch on this at the end, but to me this is an important issue that makes it hard to understand the usefulness of this step. 
- A more specific comment on the illustration using the deconfounder: while I think this is a good illustration, I think we can frame it differently because of the goals of the deconfounder in the first place. In fact, the deconfounder algorithm is literally trying recover an vector X* that renders all variables conditionally independent -- so, failing to reject that the observed number of licensed drivers is independent of the observed number of automobiles conditional on X* is more a statement that the deconfounder did not do its job to recover a good X* than it is a statement on whether they are conditionally independent given X* (X* has no real meaning, it's just a vector that's supposed to make the other variables conditionally independent). Not sure if I'm making my point clearly but let's discuss more. 

In general, my suggestion is to change the objective and tone of this section entirely, and make it about testing the results of the deconfounder. I think that's more of a contribution in and by itself, but we should discuss more.

Author: I think we're mostly together.

Absolutely +1 on testing multiple things at once, and this multiplicity diminishing the value of the test. No free lunch!

In terms of the objective of the test being to verify the deconfounder, that makes sense for the version of the test using the posterior distribution!

For the version of the distribution that uses prior distributions, I think failing to reject is a statement that the observed data don't look like the simulated data (i.e., the prior sampled X*/Y*/Z*, where X* and Y* are independent of Z*). The real usefulness of such a test is when you've created a prior such that \int P(X*|Z*)P(Z*) [dZ*] is a reasonable marginal prior for P(X) and when \int P( Y* | Z* ) P( Z* ) [dZ*]  is a reasonable prior for P(Y). At this point, the remaining discrepancies between observed and simulated data should be due to the independence assumption being violated.

I added two sentences to highlight this!

Why might this be the case?
Does travel time cause travel cost when driving alone?
Does travel cost cause travel time while driving alone?
Does some other set of variables (potentially unmeasured) cause both travel time and travel cost?

Thinking through these questions, we can immediately think of latent variables that cause both travel time, travel cost, and the choice, even after conditioning on one's travel distance.
For example, consider whether one drives alone over the San Francisco Bay Bridge.
If one crosses the bridge, then traffic delays will likely increase one's travel time.
Moreover, if one crosses the bridge, then one's travel cost is higher due to tolls that one must pay.
And finally, if one takes a toll lane to get across the bridge faster, then one also pays a higher price.
Overall, we can think of intuitive explanations for why there might be a dependence between travel time and travel cost, even after conditioning on travel distance.
These explanations suggest particular relationships to analyze and particular variables, such as bay bridge crossings, to include in one's mode choice (i.e. outcome) model.

\subsection{Additional techniques}

> \label{sec:testing-addendum}

Reviewer: I like this discussion and I think it shows exactly the right level of detail!

\section{Testing of Causal Graphs}
\label{sec:graph-testing}

\subsection{Description}
\label{sec:testing-description}

\subsubsection{Testing observable assumptions}
\label{sec:observable-tests}

> In the last section, we reviewed a process for creating an initial causal graph using expert opinion.
> Critically, after drafting a causal graph, we should immediately test it against empirical data.
> This is important because, if our graph captures inaccurate assumptions about the data generating process, then we have no reason to think that our conclusions from using the graph will be accurate.

Reviewer 2: ```suggestion
In the last section, we briefly reviewed a process for creating an initial causal graph using expert opinion.
Critically, after drafting a causal graph, we should immediately test it against available empirical data.
This is important because, if our graph captures inaccurate assumptions about the data generating process, then we have no reason to think that our conclusions from using the graph will be accurate.
```

Author: 🙏 I didn't think that section 4 was short though! ;-p

To test our causal graphs against data, we will first test the implications of our graph that involve observable variables only.
We will defer the task of testing implications that involve unobserved / latent variables to later in this subsection.
For now, recall our discussion in Section \ref{sec:graph-overview} about the two basic implications of causal graphs: marginal independence and conditional independence.
In both cases, direct testing of marginal or conditional independence amongst nodes in the causal graph may be difficult.
Indeed, there are no direct tests of conditional independence that can detect all types of dependence, especially for continuous variables \citep{bergsma_2004_testing, shah_2020_hardness}.

> As a result of this hardness, there are a myriad of research efforts aimed at testing conditional independence of two variables $X$ and $Y$, given a third variable $Z$ and any number of assumptions about the variables or the test statistic itself.

Reviewer 2: ```suggestion
As a result of this hardness, there are a myriad of recent research efforts aimed at testing conditional independence of two variables $X$ and $Y$, given a third variable $Z$ and any number of assumptions about the variables or the test statistic itself.
```

Author: Hmm, this sentence made sense to me because if something is hard, then I expect there to be many attacks at the problem (because none are completely successful), hence a myriad of research efforts.

It's def true that the papers I cite are recent research efforts. It's just not clear to me that the recency is a result of the hardness. I'm worried about your suggestion leading some readers to believe that the hardness led to the recency of efforts as opposed to the hardness leading to the many efforts.

Reviewer 2: Fair argument. We can drop it.

> Some researchers proceed under the assumption that one has access to an approximation of the conditional distribution of $X \mid Z$ \citep{candes_2018_panning, berrett_2019_conditional}.

Reviewer 2: What do you mean by proceed? Do they present a test statistic? Do they adapt an existing statistic under this assumption?

Author: The former! I changed the beginning of the sentence to "Some researchers create tests under"

Conversely, even if a set of variables are marginally or conditionally independent in their means, this \textbf{does not} imply that the variables are independent in distribution.

This approach of indirectly assessing distributional independence by testing mean independence is not new.
The following papers have all proposed and implemented such an idea: \citet{burkart_2017_predictive, chalupka_2018_fast, inacio_2019_conditional}.
For conditional independences, the crux of the approach is to predict $Y$ based on $X$ and $Z$, then compare against a prediction of $Y$ based on a resampled value of $X$ and the original $Z$.
If $Y$ is mean-independent of $X$ given $Z$, i.e. $E \left[ Y \mid X, Z \right] = E\left[ Y \mid Z \right]$, then the predictive power of a model with resampled $X$ should resemble the predictive power of a model with the original $X$.
After all, in both cases, the conditional expectation of $Y$ is independent of one's $X$ values (real or resampled).
When assessing marginal independencies, one removes $Z$ from the models for the expectation of $Y$ and proceeds as described.

> Note that as with the case of testing distributional independence, testing mean independence still requires researchers to make choices.
> We have to select models for $E \left[ Y \mid X, Z \right]$ and $E\left[ Y \mid Z \right]$, respectively, for testing conditional and marginal mean-independence.
> We also have to choose the performance statistic (e.g. $R^2$, log-likelihood, etc.) to compare these models.
> Lastly, we also have to select a resampling method.
> In particular, how (if at all) will our resampling strategy account for the possible dependence between $X$ and $Z$?

Reviewer 2: ```suggestion
Note that as with the case of testing distributional independence, testing mean independence still requires researchers to make several choices.
Firstly, we have to select models for $E \left[ Y \mid X, Z \right]$ and $E\left[ Y \mid Z \right]$, respectively, for testing conditional and marginal mean-independence.
Secondly, we also have to choose the performance statistic (e.g. $R^2$, log-likelihood, etc.) to compare these models.
Lastly, we also have to select a resampling method.
In particular, how (if at all) will our resampling strategy account for the possible dependence between $X$ and $Z$?
```

Author: 🙏

% State the choices made in this work.
% Provide some discussion of the alternatives and the effects of different choices.
For our demonstration, we made the following choices.
First, we used linear regressions to model $E \left[Y \mid X, Z \right]$ and $E \left[ Y \mid Z \right]$.

> Second, we chose to use $R^2$ as our test statistic for judging the predictive performance.

Reviewer 2: ```suggestion
Second, we chose to use $R^2$ as our test statistic for judging the predictive performance of these regressions.
```

Third, we have chosen to resample the $X$ vectors without replacement, keeping the length of the resampled vector equal to the length of the original vector.
In other words, we permute $X$.
And finally, we have chosen to visualize these tests by plotting a vertical line to display our observed test-statistic and by plotting the distribution of our test statistics based on the permutations of $Y$.

> Our rationale for these choices are as follows.

Reviewer 2: ```suggestion
Our rationales for these choices are as follows.
```

In our dataset, most of our explanatory variables were continuous (at least in theory).
Accordingly, $R^2$ seemed a sensible performance metric for a model of the conditional expectation of a continuous random variable.

In contrast to our choice of performance metric, we chose our conditional expectation models and resampling methods based on empirical testing.
In particular, we created simulations to assess our mean-independence testing procedure.
We assessed the performance of our mean-independence testing procedures using simulations where $Y \leftarrow Z \rightarrow X$ and $X$ either did or did not cause $Y$.

Of particular importance were our simulations under the null hypothesis where $X$ was conditionally independent of $Y$.
Our initial simulations used random forests as our conditional expectation models and permutations as resampling methods.

> Random forests are a non-parametric method that would allow us to have less fear of model misspecification, and permutations are easy to implement.

Reviewer 2: ```suggestion
Random forests are a non-parametric method that allowed us to have less fear of model misspecification, and permutations are easy to implement.
```

Author: Just a heads up I left this as "would allow" because we didn't use RF for the tests in the end.

However, under the null hypothesis, we discovered that the tests based on the random forest models did not result in p-values that were uniformly distributed.
Given that we planned to use these procedures in a manner akin to hypothesis-testing, we hoped that our test statistics would be U-statistics.
When we switched from the combination of random forests and permutations to linear regressions and permutations, our p-values were indeed empirically, uniformly distributed.

> Moreover, we still retained high power.

Reviewer 2: ```suggestion
However, under the null hypothesis, we discovered that the tests based on the random forest models did not result in p-values that were uniformly distributed.
Given that we planned to use these procedures in a manner akin to hypothesis-testing, we hoped that our test statistics would be U-statistics.
When we switched from the combination of random forests and permutations to linear regressions and permutations, our p-values turned out to be empirically, uniformly distributed.
Moreover, we still retained high statistical power.
```

Author: Just a heads up, I left `were indeed`. I think `turned out to be` is completely accurate. I'm just wary of what I suspect is a connotation between "turned out to be" and "these guys don't know what they're doing." I might be overly cautious about such things, but I figured the two phrases were close enough in meaning that it was better to be on (what seems to me) the safe side.

We do not claim that these choices for assessing mean independence will always be appropriate.
Indeed, one should assess one's tests on simulated data that resembles one's real data.

> For our illustrative purposes though, the combination of linear regressions, permutations, and $R^2$ resulted in adequate tests of marginal and conditional mean-independence.

Reviewer 2: What is the message here? Is it that the reader should never blindly follow any testing procedure, or is it that they should be specifically wary of what we proposed?

Author: It's def that the reader should never blindly follow any testing procedure (`assess one's tests on simulated data that resembles one's real data`).

I changed it to say 
```suggestion
We do not claim that these choices for assessing mean independence will always be appropriate.
Indeed, one should assess one's tests on simulated data that resembles one's real data.
For our dataset and simulations though, the combination of linear regressions, permutations, and $R^2$ resulted in adequate tests of marginal and conditional mean-independence.
```

Does this make it clear that I'm not telling the reader to be wary of our procedures?

Reviewer 2: Yes!
